src/main/java/gov/nasa/pds/objectAccess/TableReader.java

```diff
@@ -311,42 +315,128 @@ public int getCurrentRow() {
+
+        FileChannel inChannel = raf.getChannel();
+        int bufferSize = 1024*128;
+        if (bufferSize > inChannel.size()) {
+            bufferSize = (int) inChannel.size();
+        }
+        ByteBuffer buff = ByteBuffer.allocate(bufferSize);
+
+        byte [] bufferAsBytes = null;
+        while (inChannel.read(buff) > 0) {
+            buff = buff.position(0); // Must point the pointer to the beginning of buff indorder to access the elements in the array.
+            bufferAsBytes = buff.array();  // Get the underlying byte array in ByteBuffer.
+
+             // With the smaller buffer, we can safely read through the buffer for all lines and count them.
+             numCarriageReturns = numCarriageReturns + this.parseBufferForLineCount(dataFile, bufferAsBytes);
+
+             buff.clear();
+        }
+        raf.close();
+        return(numCarriageReturns);
+    }
+
+    private long countRecordsForTableAdapterType(URL dataFile, long offset) throws Exception {
+        // For TableCharacter, we have to rely on the size of each record and the file size to calculate the number
+        // of records thus not having to read through the entire file.
+
+        LOGGER.debug("countRecordsForTableAdapterType:dataFile,offset {},{}",dataFile,offset);
+        LOGGER.debug("countRecordsForTableAdapterType:dataFile,adapter.getRecordLength() {},{}",dataFile,adapter.getRecordLength());
+
+        long numRecords = -1;
+
+        // Do a sanity check if the record size is not known or zero.  Not all labels provide the record size, for example comma separated files.
+        // If the record size is not known or zero, unfortunately we must read through the file and count the carriage returns.
+
+        if (adapter.getRecordLength() <= 0) {
+            numRecords = this.countCarriageReturn(dataFile);
+            LOGGER.debug("countRecordsForTableAdapterType:numRecords {}",numRecords);
+            return(numRecords);
+        }
+
+        LOGGER.debug("countRecordsForTableAdapterType:numRecords:initial {}",numRecords);
+		File aFile = new File(dataFile.toURI());
+		RandomAccessFile raf = new RandomAccessFile(aFile, "r");
+		raf.seek(offset);
+	
+		FileChannel inChannel = raf.getChannel();
+		long fileSize = inChannel.size();  // The value of fileSize is now the rest content of the file after skipping any offset.
+		raf.close();
+
+        // The number of records is the size of the file divided by the record length
+        numRecords = fileSize/adapter.getRecordLength();
+
+        LOGGER.debug("countRecordsForTableAdapterType:numRecords {}",numRecords);
+
+        return(numRecords);
+    }
 	
 	/**
 	 * @return the size of record (i.e. number of lines)
 	 */
 	public long getRecordSize(URL dataFile, Object table) throws Exception {
 		adapter = AdapterFactory.INSTANCE.getTableAdapter(table);
 		InputStream is = Utility.openConnection(dataFile.openConnection());
+
+        LOGGER.debug("getRecordSize:adapter {}",adapter);
```

Comment:
@qchaupds hate to nitpick here again, but can we fix the whitespace here and elsewhere in your code updates to lineup all the tabs correctly for better readability?